### PR TITLE
Feat/pi07 paligemma

### PR DESCRIPTION
**Pull request by akshay18iitg**
## What this does
Adds the complete policy implementation for the π07 PaliGemma hierarchical architecture, consisting of a high-level planner and a low-level planner.

Low-level planner (pi07_paligemma/low_level_planner/)
V-JEPA2 video encoder replaces SigLIP — compresses each camera's temporal video into 256 tokens via a Perceiver cross-attention reducer.
Temporal state sequences (B, T, D) are projected per-timestep into separate VLM tokens (supports observation history via n_obs_history).
Subtask response conditioning — accepts an optional subtask string from the high-level planner, embedded as bidirectional tokens in the prefix.
Subgoal image conditioning — derives subgoal keys from image_features (camera{k} → subgoal{k}) following the LeRobot dataset convention; uses subgoal_is_pad to mask out dropped/absent subgoals.
Structured metadata conditioning — assembles metadata from per-sample speed, quality, mistake fields (each with _is_pad flags) into a tokenized string.
Flow matching with Knowledge Insulation — VLM KV-cache is detached before being passed to the Gemma action expert for continuous action generation.
Discrete (FAST) action prediction via the VLM backbone in parallel with continuous actions.
Prefix token layout: [video | language | response | state | subgoal images | metadata | discrete actions], with bidirectional attention for the core prefix, new bidirectional blocks for subgoals and metadata, and causal attention for discrete actions.
High-level planner (pi07_paligemma/high_level_planner/)
Uses SigLIP (via PaliGemma) for image encoding.
Autoregressive generation of updated memory tokens and subtask (response) tokens.
Structured metadata conditioning — same speed/quality/mistake assembly as low-level.
Prefix token layout: [images | language | metadata | memory | response], with bidirectional attention for images+language, a new bidirectional block for metadata, and causal attention for memory and response.
metadata_max_length added to PI07HighLevelPlannerConfig.
Configuration & types
Added metadata_max_length to both PI07lowlevelPlannerConfig and PI07HighLevelPlannerConfig.
Removed FeatureType.SUBGOAL — subgoal features are now derived from image_features at runtime.
Fixed subgoal_image_features property logic (was an impossible VISUAL and SUBGOAL check).
Docstrings
Google-style docstrings added/updated across all module, class, and method levels for both planners.
Architecture ASCII diagrams updated to reflect metadata and subgoal inputs.
## How it was tested
Integration tests for both planners (tests/policies/test_pi07_paligemma_low_level_planner.py, tests/policies/test_pi07_paligemma_high_level_planner.py) covering:
Training forward pass: verifies loss dict (MSE + CE for low-level, CE for high-level), loss finiteness, and non-empty optimizer params.
Inference pass: select_action for low-level (with action queue), sample_actions for high-level (autoregressive memory + response generation).
Pad mask verification: all token sections checked for correct ones-before-zeros pattern; image and subgoal tokens verified as never padded.
Position ID verification: monotonically incrementing for non-padded tokens, constant for padded tokens; suffix position IDs start after the cross-attention prefix (low-level).
VLM attention mask verification: full expected 2D mask reconstructed from cumsum-based attention pattern — verifies bidirectional blocks, new bidirectional blocks for subgoals/metadata, causal blocks, and correct zeroing of padded rows/columns.
Action expert attention mask verification (low-level only): cross-attention to prefix KV-cache with correct padding propagation.
Autoregressive step verification (high-level only): prefix grows by exactly 1 token per step, total steps = memory_max_length + response_max_length.
Batch data includes structured metadata fields (speed, quality, mistake + _is_pad flags), subgoal_is_pad, and subgoal0 images matching the LeRobot dataset convention.
Tests are marked @pytest.mark.gpu and @pytest.mark.slow (require CUDA).

## Checklist

- [X] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [X] My PR includes policy-related changes.
  - [x] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### CODE REVIEW

File: src/opentau/policies/pi07_paligemma/high_level_planner/modeling_pi07_high_level.py [hunk 1]
```diff
@@ -0,0 +1,1361 @@
...
+
+        metadata = []
+        for speed, quality, mistake, speed_is_pad, quality_is_pad, mistake_is_pad in zip(
+            batch["speed"],
+            batch["quality"],
+            batch["mistake"],
+            batch["speed_is_pad"],
+            batch["quality_is_pad"],
+            batch["mistake_is_pad"],
+            strict=False,
```
Comment (shuheng-liu) [inline src/opentau/policies/pi07_paligemma/high_level_planner/modeling_pi07_high_level.py]: Nit: `strict = True` is safer

File: src/opentau/policies/pi07_paligemma/high_level_planner/modeling_pi07_high_level.py [hunk 2]
```diff
@@ -0,0 +1,1361 @@
...
+            if not speed_is_pad:
+                meta += f"Speed: {str(speed.item())} "
+
+            if not quality_is_pad:
+                meta += f"Quality: {str(quality.item())} "
+
+            if not mistake_is_pad:
+                meta += f"Mistake: {str(mistake.item())}"
+
+            metadata.append(f"Metadata: {meta}<eos>")
```
Comment (shuheng-liu) [inline src/opentau/policies/pi07_paligemma/high_level_planner/modeling_pi07_high_level.py]: Nit: there might be a spurious whitespace at the end if "mistake" is absent and speed/quality is present. I'd build a list of segments and use a `" ".join(...)` instead

**Review by shuheng-liu — DISMISSED** [review]

**Review by shuheng-liu — DISMISSED** [review]

**Review by shuheng-liu — APPROVED** [review]
